gadget/partial.go

```diff
@@ -0,0 +1,129 @@
+// properties partially defined. After that it checks that the gadget
+// is now fully specified.
+func ApplyInstallerVolumesToGadget(installerVols map[string]*Volume, lovs map[string]*LaidOutVolume) error {
+	for volName, lov := range lovs {
+		if len(lov.Partial) == 0 {
+			continue
+		}
+
+		insVol := installerVols[volName]
+		if insVol == nil {
+			return fmt.Errorf("installer did not provide information for volume %q", volName)
+		}
+
+		// TODO: partial structure, as it is not clear what will be possible when set
+
+		if lov.HasPartial(PartialSchema) {
+			if insVol.Schema == "" {
+				return fmt.Errorf("installer did not provide schema for volume %q", volName)
+			}
+			lov.Schema = insVol.Schema
+		}
+
+		if lov.HasPartial(PartialFilesystem) {
+			if err := applyPartialFilesystem(insVol, lov, volName); err != nil {
+				return err
+			}
+		}
+
+		if lov.HasPartial(PartialSize) {
+			if err := applyPartialSize(insVol, lov, volName); err != nil {
+				return err
+			}
+		}
+
+		// The only thing that can still be partial is the structure
+		if lov.HasPartial(PartialStructure) {
+			lov.Partial = []PartialProperty{PartialStructure}
+		} else {
+			lov.Partial = []PartialProperty{}
+		}
+
+		// Now validate finalized volume
+		if err := validateVolume(lov.Volume); err != nil {
+			return fmt.Errorf("finalized volume %q is wrong: %v", lov.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func applyPartialFilesystem(insVol *Volume, lov *LaidOutVolume, volName string) error {
+	for sidx := range lov.Structure {
+		// Two structures to modify due to copies inside LaidOutVolume
+		vs := &lov.Structure[sidx]
+		vsLos := lov.LaidOutStructure[sidx].VolumeStructure
+		insStr, err := structureByName(insVol.Structure, vs.Name)
+		if err != nil {
+			return err
+		}
+		if vs.Filesystem == "" && vs.HasFilesystem() {
+			if insStr.Filesystem == "" {
+				return fmt.Errorf("installer did not provide filesystem for structure %q in volume %q", vs.Name, volName)
+			}
+			vs.Filesystem = insStr.Filesystem
+			vsLos.Filesystem = insStr.Filesystem
+		}
+	}
+	return nil
+}
+
+func applyPartialSize(insVol *Volume, lov *LaidOutVolume, volName string) error {
+	for sidx := range lov.Structure {
+		// Two structures to modify due to copies inside LaidOutVolume
+		vs := &lov.Structure[sidx]
+		vsLos := lov.LaidOutStructure[sidx].VolumeStructure
+		insStr, err := structureByName(insVol.Structure, vs.Name)
+		if err != nil {
+			return err
+		}
+		if vs.hasPartialSize() {
+			if insStr.Size == 0 {
+				return fmt.Errorf("installer did not provide size for structure %q in volume %q", vs.Name, volName)
+			}
+			if insStr.Offset == nil {
+				return fmt.Errorf("installer did not provide offset for structure %q in volume %q", vs.Name, volName)
+			}
+			vs.Size = insStr.Size
+			vsLos.Size = insStr.Size
+			vs.Offset = insStr.Offset
+			vsLos.Offset = insStr.Offset
+			lov.LaidOutStructure[sidx].StartOffset = *insStr.Offset
+		}
```

Comment:
(personal preference) again here it makes sense to reverse the logic checks
```suggestion
		if !vs.hasPartialSize() {
		    continue
		}
	        if insStr.Size == 0 {
		        return fmt.Errorf("installer did not provide size for structure %q in volume %q", vs.Name, volName)
	        }
	        if insStr.Offset == nil {
		        return fmt.Errorf("installer did not provide offset for structure %q in volume %q", vs.Name, volName)
	        }
	        vs.Size = insStr.Size
	        vsLos.Size = insStr.Size
	        vs.Offset = insStr.Offset
	        vsLos.Offset = insStr.Offset
	        lov.LaidOutStructure[sidx].StartOffset = *insStr.Offset
```